server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/BlobPutter.java

```diff
@@ -38,7 +42,45 @@
 
 public interface BlobPutter {
 
+    class TempBlob {
+
+        private static void destroy(TempBlob tempBlob) {
+            FileUtils.deleteQuietly(tempBlob.content);
+        }
+
+        private final BlobId blobId;
+        private final File content;
+
+        private TempBlob(BlobId blobId, File content) {
+            this.blobId = blobId;
+            this.content = content;
+        }
+
+        public BlobId getBlobId() {
+            return blobId;
+        }
+
+        public File getContent() {
+            return content;
+        }
+    }
+
     Mono<Void> putDirectly(ObjectStorageBucketName bucketName, Blob blob);
 
-    Mono<BlobId> putAndComputeId(ObjectStorageBucketName bucketName, Blob initialBlob, Supplier<BlobId> blobIdSupplier);
+    default Mono<BlobId> putAndComputeId(ObjectStorageBucketName bucketName, Blob initialBlob, Supplier<BlobId> blobIdSupplier) {
+        return Mono.using(
+            () -> temporallySaving(initialBlob, blobIdSupplier),
+            tempBlob -> putTempBlob(bucketName, tempBlob)
+                .then(Mono.just(tempBlob.blobId)),
+            TempBlob::destroy);
+    }
+
+    default TempBlob temporallySaving(Blob blob, Supplier<BlobId> blobIdSupplier) throws IOException {
```

Comment:
with this code, the file saving is synchronous, it's bad for performances I guess

Reply:
The old fashion uploading is still doing it synchronously, there is one possible explanation I can think about is writing blob into file then into swift is possible slower than writing directly into swift then copy + remove the temporally (which can be blazing fast)